components/calendar/src/chinese_based.rs

```diff
@@ -0,0 +1,346 @@
+//! assert_eq!(chinese_date.month().ordinal, 6);
+//! assert_eq!(chinese_date.day_of_month().0, 6);
+//! ```
+
+use crate::{
+    astronomy::{Astronomical, Location, MEAN_SYNODIC_MONTH, MEAN_TROPICAL_YEAR},
+    calendar_arithmetic::{ArithmeticDate, CalendarArithmetic},
+    helpers::{adjusted_rem_euclid, i64_to_i32, quotient, I32Result},
+    rata_die::RataDie,
+    types::Moment,
+    Calendar, Date,
+};
+
+/// The trait ChineseBased is used by Chinese-based calendars to perform computations shared by such calendar.
+/// To do so, calendars should:
+///
+/// - Implement `fn location` by providing a location at which observations of the moon are recorded, which
+/// may change over time (the zone is important, long, lat, and elevation are not relevant for these calculations)
+/// - Define `const EPOCH` as a `RataDie` marking the start date of the era of the Calendar for internal use,
+/// which may not accurately reflect how years or eras are marked traditionally or seen by end-users
+/// - Implement `fn new_chinese_based_date` by taking a year, month, and day in a Chinese-based calendar and
+/// returning a Date of the relevant Calendar type.
+///
+/// For an example of how to use this trait, see `impl ChineseBased for Chinese` in [`Chinese`].
+pub(crate) trait ChineseBased: CalendarArithmetic + Sized {
+    /// Given a fixed date, return the location used for observations of the new moon in order to
+    /// calculate the beginning of months. For multiple Chinese-based lunar calendars, this has
+    /// changed over the years, and can cause differences in calendar date.
+    fn location(fixed: RataDie) -> Location;
+
+    /// The RataDie of the beginning of the epoch used for internal computation; this may not
+    /// reflect traditional methods of year-tracking or eras, since Chinese-based calendars
+    /// may not track years ordinally in the same way many western calendars do.
+    const EPOCH: RataDie;
+
+    /// Given a year, month, and day, create a Date<C> where C is a Chinese-based calendar.
+    ///
+    /// This function should just call try_new_C_date where C is the name of the calendar.
+    fn new_chinese_based_date(year: i32, month: u8, day: u8) -> Date<Self>;
+}
+
+/// Chinese-based calendars define DateInner as a calendar-specific struct wrapping ChineseBasedDateInner.
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq, PartialOrd, Ord)]
+pub(crate) struct ChineseBasedDateInner<C: ChineseBased + CalendarArithmetic>(
+    pub(crate) ArithmeticDate<C>,
+);
+
+impl<C: ChineseBased + CalendarArithmetic> ChineseBasedDateInner<C> {
+    /// Get the current major solar term of a fixed date, output as an integer from 1..=12.
+    ///
+    /// Based on functions from _Calendrical Calculations_ by Reingold & Dershowitz.
+    /// Lisp reference code: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L5273-L5281
+    pub(crate) fn major_solar_term_from_fixed(date: RataDie) -> i32 {
```

Comment:
Question: if this is 1..12 then can we return u8 instead? Or maybe a u32, if you make adjusted_rem_euclid return u32 or similar. Could be in a follow-up.

Reply:
I've added a TODO and put this as a task in #3661 to address in my follow-up PR